File: algorithm/leetcode/92_Reverse_Linked_List_II.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} m
 * @param {number} n
 * @return {ListNode}
 */
var reverseBetween = function(head, m, n) {
    let dummy = new ListNode(-1);
    dummy.next = head
    head = dummy
    for(let i = 1; i< m; i++) {
        head = head.next
    }
    let headOfSubList = head.next,
        nodeBeforeHead = head,
        nextNode = head.next.next,
        currNode = head.next;
    for(let i = m; i < n; i++) {
        let temp = nextNode.next;
        nextNode.next = currNode;
        currNode = nextNode;
        nextNode = temp
    }
    headOfSubList.next = nextNode;
    nodeBeforeHead.next = currNode;
    return dummy.next
};

File: algorithm/leetcode/92_Reverse_Linked_List_II.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode reverseBetween(ListNode head, int m, int n) {
        ListNode dumy = new ListNode(0);
        dumy.next = head;
        head = dumy;
        for(int i = 1; i < m; i++) {
            head = head.next;
        }
        ListNode headOfSubList = head.next;
        ListNode nodeBeforeList = head;
        ListNode nextNode = head.next.next;
        ListNode currNode = head.next;
        for(int i = m; i < n; i++) {
            ListNode temp = nextNode.next;
            nextNode.next = currNode;
            currNode = nextNode;
            nextNode = temp;
        }
        nodeBeforeList.next = currNode;
        headOfSubList.next = nextNode;
        return dumy.next;
    }
}